llvm/lib/MC/GOFFObjectWriter.cpp

```diff
@@ -216,34 +220,471 @@ void GOFFOstream::write_impl(const char *Ptr, size_t Size) {
+      PR->NameSpace = GOFF::ESD_NS_Parts;
+
+      GSection = createGOFFSection(PR, PR, SD, &Section);
+    }
+  } else
+    llvm_unreachable("Unhandled section kind");
+
+  SectionMap[&Section] = GSection;
+}
+
+void GOFFObjectWriter::processSymbolDefinedInModule(
+    const MCSymbolGOFF &MCSymbol, const MCAssembler &Asm,
+    const MCAsmLayout &Layout) {
+  MCSection &Section = MCSymbol.getSection();
+  SectionKind Kind = Section.getKind();
+  auto &Sec = cast<MCSectionGOFF>(Section);
+
+  GOFFSection *GSection = SectionMap[&Sec];
+  assert(GSection && "No corresponding section found");
+
+  GOFFSymbol *GSectionSym = GSection->Pptr;
+  assert(GSectionSym &&
+         "Defined symbols must exist in an initialized GSection");
+
+  StringRef SymbolName = MCSymbol.getName();
+  // If it's a text section, then create a label for it.
+  if (Kind.isText()) {
+    GOFFSymbol *LD = createLDSymbol(SymbolName, GSectionSym->EsdId);
+    LD->BindingStrength = MCSymbol.isWeak()
+                              ? GOFF::ESDBindingStrength::ESD_BST_Weak
+                              : GOFF::ESDBindingStrength::ESD_BST_Strong;
+
+    // If we don't know if it is code or data, assume it is code.
+    LD->Executable = MCSymbol.getExecutable();
+    if (LD->isExecUnspecified())
+      LD->Executable = GOFF::ESD_EXE_CODE;
+
+    // Determine the binding scope. Please note that the combination
+    // !isExternal && isExported makes no sense.
+    LD->BindingScope = MCSymbol.isExternal()
+                           ? (MCSymbol.isExported()
+                                  ? GOFF::ESD_BSC_ImportExport
+                                  : (LD->isExecutable() ? GOFF::ESD_BSC_Library
+                                                        : GOFF::ESD_BSC_Module))
+                           : GOFF::ESD_BSC_Section;
+
+    if (ADA && ADA->SectionLength > 0)
+      LD->ADAEsdId = ADA->EsdId;
+    else
+      LD->ADAEsdId = 0;
+
+    GSectionSym->setMaxAlignment(MCSymbol.getAlignment());
+
+    LD->MCSym = &MCSymbol;
+    SymbolMap[&MCSymbol] = LD;
+  } else if (Kind.isBSS() || Kind.isData()) {
+    GSectionSym = GSection->Rptr;
+    GSectionSym->BindingScope =
+        MCSymbol.isExternal()
+            ? (MCSymbol.isExported() ? GOFF::ESD_BSC_ImportExport
+                                     : GOFF::ESD_BSC_Library)
+            : GOFF::ESD_BSC_Section;
+    if (GSectionSym->BindingScope == GOFF::ESD_BSC_Section)
+      GSection->SD->BindingScope = GOFF::ESD_BSC_Section;
+    GSectionSym->setAlignment(MCSymbol.getAlignment());
+    GSectionSym->SectionLength = Layout.getSectionAddressSize(&Section);
+
+    // We cannot have a zero-length section for data.  If we do, artificially
+    // inflate it.  Use 2 bytes to avoid odd alignments.
+    if (!GSectionSym->SectionLength)
+      GSectionSym->SectionLength = 2;
+
+    GSectionSym->MCSym = &MCSymbol;
+    GSection->SD->MCSym = &MCSymbol;
+    SymbolMap[&MCSymbol] = GSectionSym;
```

Comment:
In GOFF, are there any possibility that a MCSection have several different MCSymbols ?  if there are, 
 different MCSymbols(in the same MCSection ) in SymbolMap[MCSymbol] will has the same GSectionSym?

Reply:
GSectionSym should be GOFFSymbol (or rather pointer to GOFFSymbol). 

It represents the ESD Entry that "owns" the text/data that is the section in question. 

Do you think the name `GSectionSym` is unclear? If so I can change the name to somehting better.